### DISCUSSION TITLE
How to Receive Bang & Olufsen Repeat Codes

### DISCUSSION THREAD

**Discussion by bm7000** [Arduino-IRremote/Arduino-IRremote#1181] [Q&A] [answered]
Hello Experts:

I am trying to write an Arduino sketch for receiving/reading B&O remote commands from a Beo4 remote control.
I am using a TSOP7000 receiver connected to pin 10. It seems to work well, and I am receiving codes when I press buttons on the remote. 

What mystifies me is: 

How to read out/detect that the Beo4 sends repeat codes when a button is pressed continuously. I know that this only works with some of the buttons like >> and << and up/down and volume. I can see that the remote is sending codes when these buttons are pressed continuously from a connected oscilloscope and also from the 'LED feedback' that is enabled in my sketch. In other words, the LED is blinking fast, but no new codes are produced when IrReceiver.decode() is called. The 'flags' also do not change.

This is my sketch:

#define DECODE_BEO //turns on B&O decoding in IRremote.hpp
#define RECORD_GAP_MICROS 20000 //the prescribed minimum of 16000 did not work...this command is necessary to compile the sketch.

#include <IRremote.hpp>

int IR_RECEIVE_PIN = 10;

void setup()
{
  Serial.begin(9600);
  
  IrReceiver.begin(IR_RECEIVE_PIN, ENABLE_LED_FEEDBACK); // Start the receiver
}

void loop() {
  if (IrReceiver.decode()) {//has a transmission been received?
   
    Serial.print("protocol ");Serial.println(IrReceiver.decodedIRData.protocol);
    Serial.print("address ");Serial.println(IrReceiver.decodedIRData.address);
    Serial.print("extra ");Serial.println(IrReceiver.decodedIRData.extra);
    Serial.print("command ");Serial.println(IrReceiver.decodedIRData.command);
    Serial.print("decodedRawData ");Serial.println(IrReceiver.decodedIRData.decodedRawData,HEX);
    Serial.print("flags ");Serial.println(IrReceiver.decodedIRData.flags);
    Serial.println();
        
    IrReceiver.resume(); // Receive the next value
  }
}

A typical output looks like (Beo4 set to "TV" and volume up is pressed):

protocol 24
address 0
extra 0
command 96
decodedRawData 60
flags 128

Thanks much in advance for your help! And my apologies if my questions has an obvious solution that escaped me. I am a bit out of my depth here...;-).

**Comment by ArminJo** [comment]
Please post the serial output of the SimpleReceiver example for a repeated and a not repeated command.
Maybe we can evaluate the gap between the repeats to set the repeat flag, like for other protocols.

**Comment by bm7000** [comment]
Thanks much for your reply!

I tried the SimpleReceiver, and it does the same. The LED blinks rapidly when the remote sends repeat codes, but there are no additional printouts. Just this (pressing the volume up button on the Beo4):

Protocol=Bang&Olufsen Address=0x1 Command=0x60 Raw-Data=0x160 16 bits MSB first
Send with: IrSender.sendBang&Olufsen(0x1, 0x60, <numberOfRepeats>);

**Comment by bm7000** [comment]
Learning process: I used 9600 baud for the Serial port in the above test. I switched back to the 115200 in the original SimpleReceiver sketch and now I get this:

Protocol=Bang&Olufsen Address=0x1 Command=0x60 Raw-Data=0x160 16 bits MSB first
Send with: IrSender.sendBang&Olufsen(0x1, 0x60, <numberOfRepeats>);

Overflow
Received noise or an unknown (or not yet enabled) protocol
rawData[112]: 
 -33800
 + 250,-2850
 + 200,-2900 + 200,-15250 + 200,-2900 + 200,-5950
 + 250,-5950 + 200,-6000 + 200,-5950 + 250,-5950
 + 200,-6000 + 200,-5950 + 250,-9050 + 200,-2900
 + 200,-9050 + 250,-5950 + 200,-2900 + 200,-5950
 + 250,-5950 + 200,-5950 + 250,-5950 + 200,-12200
 + 200,-2850 + 250,-2850 + 250,-15200 + 250,-2850
 + 200,-6000 + 200,-5950 + 250,-5950 + 250,-5950
 + 200,-5950 + 250,-5950 + 200,-6000 + 200,-9050
 + 250,-2850 + 200,-9050 + 250,-5950 + 250,-2850
 + 200,-6000 + 200,-5950 + 250,-5950 + 200,-5950
 + 250,-12150 + 200,-2900 + 200,-2850 + 250,-15250
 + 200,-2900 + 200,-5950 + 250,-5950 + 200,-5950
 + 250,-5950 + 200,-6000 + 200,-5950 + 250,-5950
 + 250,-9050 + 200,-2850 + 250
Sum: 352650

**Reply by ArminJo** [reply]
You read [this](https://github.com/Arduino-IRremote/Arduino-IRremote/blob/8c84e03318bf219fdbba910642acba87989d58d4/src/ir_BangOlufsen.hpp#L68)?

**Comment by bm7000** [comment]
Yes, but I guess I did not read it throughly enough. I found what you marked: "Can only receive single messages and back to back repeats will result in overflow"
I have to admit I do not fully understand the statements in point 2.) of this comment. I just added #define RECORD_GAP_MICROS to my sketch since otherwise it would not compile. Bottom line: Is there a way to get the repeat codes read out or is this just a problem to work around, for example by reading out the feedback LED pin state and use that together with the last received command to detect repeats?

**Reply by ArminJo** [reply, marked as answer]
Did you do this?
#define ENABLE_BEO_WITHOUT_FRAME_GAP 
#define RECORD_GAP_MICROS 13000

**Comment by bm7000** [comment]
I just tried. And it is working! Awesome! Thank you very much!

Could you explain what the 'Frame Gap' has to do with this?
